File: 03. LeetCode/121. Best Time to Buy and Sell Stock/Solution.js

var maxProfit = function(prices) {
    let buy = 10000;
    let answer = 0;
    
    for(let i = 0; i < prices.length; i++){
        if(prices[i] < buy){
            buy = prices[i];
        }else{
            answer = Math.max(answer, prices[i] - buy);
        }
    }
    return answer;
};

console.log( maxProfit([7, 1, 5, 3, 6, 4]) );
console.log( maxProfit([7, 6, 4, 3, 1]) );

File: 03. LeetCode/121. Best Time to Buy and Sell Stock/Solution.java
import java.util.*;

class Solution {
    
    public int maxProfit(int[] prices) {
        int answer = 0;
        int buy = Integer.MAX_VALUE;
        
        for(int i = 0; i < prices.length; i++){
            if(prices[i] < buy){
                buy = prices[i];
            }else{                
                answer = Math.max(answer, prices[i] - buy);
            }            
        }
        return answer;
    }

	public static void main(String[] args) {
		Solution solution = new Solution();
		int[] prices = { 7, 1, 5, 3, 6, 4 };
		System.out.println( solution.maxProfit(prices) );
		
		int[] prices2 = { 7, 6, 4, 3, 1 };
		System.out.println( solution.maxProfit(prices2) );
	}
}